go/pkg/client/cas_upload.go

```diff
@@ -127,11 +131,18 @@ func (c *Client) BatchWriteBlobs(ctx context.Context, blobs map[digest.Digest][]
 	var reqs []*repb.BatchUpdateBlobsRequest_Request
 	var sz int64
 	for k, b := range blobs {
-		sz += int64(k.Size)
-		reqs = append(reqs, &repb.BatchUpdateBlobsRequest_Request{
+		r := &repb.BatchUpdateBlobsRequest_Request{
 			Digest: k.ToProto(),
 			Data:   b,
-		})
+		}
+		if c.batchCompression && c.shouldCompress(k.Size) {
+			r.Data = zstdEncoder.EncodeAll(r.Data, nil)
+			r.Compressor = repb.Compressor_ZSTD
+			sz += int64(len(r.Data))
+		} else {
+			sz += int64(k.Size)
```

Comment:
It's tempting to use `len(r.Data)` here as well, but this keeps this path identical to what it was before this change. Good call!